lib/graphql/schema/reduce_types.rb

```diff
@@ -28,32 +36,46 @@ def self.reduce_type(type, type_hash, context_description)
         if !type_hash.fetch(type.name, nil).equal?(type)
           validate_type(type, context_description)
           type_hash[type.name] = type
-          crawl_type(type, type_hash, context_description)
+          crawl_type(type, type_hash, context_description, renamer)
         end
       end
 
-      def self.crawl_type(type, type_hash, context_description)
+      def self.crawl_type(type, type_hash, context_description, renamer)
         if type.kind.fields?
-          type.all_fields.each do |field|
-            reduce_type(field.type, type_hash, "Field #{type.name}.#{field.name}")
-            field.arguments.each do |name, argument|
-              reduce_type(argument.type, type_hash, "Argument #{name} on #{type.name}.#{field.name}")
+          type.fields.keys.each do |name|
+            original_field = type.fields.delete(name)
+            field = renamer.rename_field(original_field)
+
+            reduce_type(field.type, type_hash, "Field #{type.name}.#{field.name}", renamer)
+
+            field.arguments.keys.each do |argument_name|
+              original_argument = field.arguments.delete(argument_name)
+              argument = renamer.rename_argument(original_argument)
+
+              reduce_type(argument.type, type_hash, "Argument #{name} on #{type.name}.#{field.name}", renamer)
+              field.arguments[argument.name] = argument
             end
+
+            type.fields[field.name] = field
           end
         end
         if type.kind.object?
           type.interfaces.each do |interface|
-            reduce_type(interface, type_hash, "Interface on #{type.name}")
+            reduce_type(interface, type_hash, "Interface on #{type.name}", renamer)
           end
         end
         if type.kind.union?
           type.possible_types.each do |possible_type|
-            reduce_type(possible_type, type_hash, "Possible type for #{type.name}")
+            reduce_type(possible_type, type_hash, "Possible type for #{type.name}", renamer)
           end
         end
         if type.kind.input_object?
-          type.arguments.each do |argument_name, argument|
-            reduce_type(argument.type, type_hash, "Input field #{type.name}.#{argument_name}")
+          type.arguments.keys.each do |argument_name|
+            original_argument = type.arguments.delete(argument_name)
+            argument = renamer.rename_argument(original_argument)
+
+            reduce_type(argument.type, type_hash, "Input field #{type.name}.#{argument_name}", renamer)
+            type.arguments[argument.name] = argument
```

Comment:
Could probably save a little bit of work (delete and then re-insert the same thing) here by checking if the renamer isn't provided.